### PR TITLE
Add tform to tps

### PR DESCRIPTION
**Pull request by djkapner**
Submitting PR for first codecov and to get comments:
Has tform().
Took a shot at moving this transform into its own folder... soliciting comments on how I did that.
Test code remains to be written...

### CODE REVIEW

**Comment by RussTorres** [comment]
I think it's cleaner to have the `transform` module be put in a directory e.g.
```
renderapi/
  transform/
    __init__.py (imports * from transform.py, tps.py, etc. for compatibility)
    transform.py (as is, or with e.g. AffineModel instances pulled out)
    thinplatespline.py
```

I'm not sure this exact setup would avoid circular imports with something like `load_leaf_transform`, but it would be straightforward to debug that and put things where they belong.

**Comment by djkapner** [comment]
With this reorg, I'm having some trouble with the conditional scipy/numpy import statement in transform.py. I'm not quite sure how that is working, especially in test/test_transform/test_Polynomial_estimation

**Comment by djkapner** [comment]
I'm also wondering if this is a legacy switch, since according to this, svd and LinAlgError are the same between the two:
https://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html#scipy-linalg-vs-numpy-linalg

**Comment by djkapner** [comment]
In my experience, the best backend to use (speed) is MKL, and within a conda environment `conda install numpy` will give an MKL backend, while `pip install numpy` will not. It seems like we could get away from scipy entirely?

**Comment by RussTorres** [comment]
Where do you see that `scipy.linalg.svd` and `numpy.linalg.svd` are the same?  I'm pretty sure this has not been the case since this repo started -- they take completely different inputs.

**Comment by RussTorres** [comment]
In the short term, have you tried reloading and checking against `renderapi.transform.transform`?

**Comment by djkapner** [comment]
This comment led me to believe they are the same:
["scipy.linalg contains all the functions in numpy.linalg. plus some other more advanced ones not contained in numpy.linalg"](https://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html#scipy-linalg-vs-numpy-linalg)
but, you're right, scipy seems to have some additional knobs:
`scipy.linalg.svd(a, full_matrices=True, compute_uv=True, overwrite_a=False, check_finite=True, lapack_driver='gesdd')`
`numpy.linalg.svd(a, full_matrices=1, compute_uv=1)`
given how render-python is calling svd, these differences are not exploited:
```
danielk@ibs-danielk-ux1:~/djk_render-python$ grep -rnw './' -e 'svd'
./build/lib/renderapi/transform.py:20:    from scipy.linalg import svd, LinAlgError
./build/lib/renderapi/transform.py:25:    from numpy.linalg import svd
./build/lib/renderapi/transform.py:918:        U, S, V = svd(A)
./build/lib/renderapi/transform.py:1161:        _, s, V = svd(A)
./renderapi/transform/transform.py:20:    from scipy.linalg import svd, LinAlgError
./renderapi/transform/transform.py:25:    from numpy.linalg import svd
./renderapi/transform/transform.py:918:        U, S, V = svd(A)
./renderapi/transform/transform.py:1161:        _, s, V = svd(A)
./test/test_transform.py:151:    assert(renderapi.transform.svd is np.linalg.svd
./test/test_transform.py:152:           if use_numpy else renderapi.transform.svd is scipy.linalg.svd)
./test/test_transform.py:176:    assert(renderapi.transform.svd is scipy.linalg.svd)
```

**Comment by djkapner** [comment]
re: reloading
I'm not sure what you mean here. I'm trying to address this [failure](https://travis-ci.org/fcollman/render-python/jobs/413184013)

**Comment by fcollman** [comment]
just a general comment... you should use the __all__ keyword to specify explicitly what import * means

**Comment by djkapner** [comment]
I'm looking at adding inverse_tform() to the TPS. The render/trakem2/jitk inverse doesn't actually seem to work much at all, but, what they have in there is simply a gradient descent solver to brute force out the inverse point. In python, I can do something like this
```
    def inverse_optfun(self, pt, pt0):
        return self.apply(pt) - pt0

    def inverse_tform(self, points):
        """transform a set of points through the inverse of this transformation
        Parameters
        ----------
        points : numpy.array
            a Nx2 array of x,y points
        Returns
        -------
        numpy.array
            a Nx2 array of x,y points after inverse transformation
        """
        newpts = []
        for p in points:
            npt = root(self.inverse_optfun, p, args=(p,))
            newpts.append(npt.x)
        return np.array(newpts)
```
but that requires scipy. Is there a way to offload one class function with scipy dependencies to external?
I could also write a numpy-based gradient descent into it, but, ...

**Comment by djkapner** [comment]
I'm not crazy about how slow and brutish this inverse_tform() is, but, ... potentially we could move this method and gradient_descent into the general transform and call it inverse_tform_bruteforce() which should then work with other non-analytic transforms...

**Comment by fcollman** [comment]
I think we should try to maintain a uniform interface... if we come up with a better implementation later that's fine, but the api should be fixed. we probably should move to merge this sooner rather than later and not let it get mission creep.  It already has a re-org and a new tps implementation rolled into one.

File: renderapi/transform/leaf/ThinPlateSpline.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+import numpy as np
```
Comment (fcollman) [inline renderapi/transform/leaf/ThinPlateSpline.py]: module names should be all lowercase with _ not CamelCase

File: renderapi/transform/leaf/PolynomialModels.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
+from .transform import Transform, logger
```
Comment (fcollman) [inline renderapi/transform/leaf/PolynomialModels.py]: this is another example of a CamelCase module name.. should be lower case with _

File: renderapi/transform/leaf/AffineModels.py [hunk 3]
```diff
@@ -0,0 +1,628 @@
+from .transform import Transform, logger
```
Comment (fcollman) [inline renderapi/transform/leaf/AffineModels.py]: CamelCase problem...

**Comment by djkapner** [comment]
it still remains to add fit() and estimate() to ThinPlateSplineTransform but I can come back to that later in a different PR

**Comment by fcollman** [comment]
yeah I think that sounds like a separate PR

File: renderapi/transform/utils.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
+from collections import Iterable
+from renderapi.errors import RenderError
+from .leaf import AffineModel, Polynomial2DTransform
+from .transform import TransformList, ReferenceTransform
+
+
```
Comment (fcollman) [inline renderapi/transform/utils.py]: this is where the __all__ = ['estimate_dstpts', 'estimate_transformsum'] should be (and by analogy everywhere else that you want the ability to do an import *

**Comment by RussTorres** [comment]
@fcollman @djkapner are we ready to merge this PR?

**Comment by djkapner** [comment]
I think so. I'll make an issue to add fit() and estimate() later.

**Pull request merged**